### DISCUSSION TITLE
Finding minimum weight perfect matching with nx.min_weight_matching(G)

### DISCUSSION THREAD

**Discussion by jeffgong047** [networkx/networkx#7513] [Q&A]
Hi Developers,
I encountered an issue with using nx.min_weight_matching() for finding minimum weight perfect matching when the graph has negative weights. I am pretty sure that the presented graph contains a perfect matching that has minimum weight and I have at least one simple example where the minimum weight perfect matching is not found by the algorithm as indicated in the attached figure (the red edges are matched, however the MWPM should include both negative weighted edges). Could you let me know whether this implementation could handle the case with negative weights and requirement for a perfect matching. 

![expandedG_dual_expanded_2_2](https://github.com/networkx/networkx/assets/66967298/9cd39418-c563-452d-860d-019e86ecac0b)

**Comment by rossbar** [comment]
> I am pretty sure that the presented graph contains a perfect matching that has minimum weight and I have at least one simple example where the minimum weight perfect matching is not found by the algorithm as indicated in the attached figure (the red edges are matched, however the MWPM should include both negative weighted edges).

Are you sure? At first glance[^1], it appears that using any one of the "central" edges (i.e. the four connecting to the central pod with non-zero weights) will require using all four of them... since the sum of these four edges is > 0, then that would not be a minimum matching. A quick way to check this would be to modify one of those edge weights so that the sum is < 0.

> Could you let me know whether this implementation could handle the case with negative weights and requirement for a perfect matching.

While the docstring doesn't mention negative weights explicitly, it *does* [highlight how weights are handled](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.matching.min_weight_matching.html#min-weight-matching) - quoting the docstring:

> This method replaces the edge weights with 1 plus the maximum edge weight minus the original edge weight.
> 
> new_weight = (max_weight + 1) - edge_weight
> 
> then runs [max_weight_matching()](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.matching.max_weight_matching.html#networkx.algorithms.matching.max_weight_matching) > with the new weights.

Given this info, do you expect the algorithm to perform properly on your data?

[^1]: This is just based on intuition, I didn't actually attempt to evaluate the problem so it may be wrong!

**Reply by jeffgong047** [reply]
Thank you so much for quick reply! The explanation of using max_weight_matching for solving min_weight_matching with weight conversion is quite clear, however I still could not get MWPM on small examples. In the attachment is one old example and a new example, and the red lines is the outcome of the algorithm, where the blue lines are manually selected by me which should be the ground truth. To summarize, this problem only appears when the graph contains negative weights, so my hypothesis is that the implementation provided for minimum weight matching can not deal with negative weights or negative weights with perfect matching requirement. 
![expandedG_dual_expanded_2_2_negative](https://github.com/networkx/networkx/assets/66967298/2d76e59c-c203-439c-9ad2-ca58af21ebfa)
![expandedG_dual_expanded_2_2](https://github.com/networkx/networkx/assets/66967298/73fa732e-f246-4e56-b5bd-f14b6604d9a5)

**Comment by rossbar** [comment]
Yes, the example with all negative central weights is compelling - those edges would certainly be expected to be included in the minimal weight matching.

I suspect there's an implicit assumption of non-negative weights in the implementation (which is not currently documented). The obvious thing to try would be to modify the formula to make all of the weights non-negative, e.g. by determining the minimum weight and adding its absolute value to `new_weight` if it happens to be negative.

**Reply by jeffgong047** [reply]
Thanks for the feedback! I will consider that if transformation works and probably some other algorithm.
